src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,162 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null && tail == null) {
+            head = new Node(null, value, null);
+            tail = head;
+            size++;
+        } else {
+            head.next = new Node<>(head, value, null);
+            head = head.next;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0 && head == null && tail == null) {
+            head = new Node(null, value, null);
+            tail = head;
+        } else if (index == 0) {
+            Node<T> nodeToAdd = new Node<>(null, value, tail);
+            tail.prev = nodeToAdd;
+            tail = nodeToAdd;
+        } else if (index == size) {
+            add(value);
+            size--;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> nodeToAdd = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = nodeToAdd;
+            currentNode.prev = nodeToAdd;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el : list) {
+            add(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode = getNode(index);
+        Node<T> nodeToSet = new Node<>(currentNode.prev, value, currentNode.next);
+        if (currentNode.next != null) {
+            currentNode.next.prev = nodeToSet;
+        } else {
+            head = nodeToSet;
+        }
+        if (currentNode.prev != null) {
+            currentNode.prev.next = nodeToSet;
+        } else {
+            tail = nodeToSet;
+        }
+        return currentNode.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode = getNode(index);
+
+        remove(currentNode.item);
```

Comment:
The `remove(int index)` method should directly unlink the node instead of calling `remove(currentNode.item)`. Consider creating a separate method `unlink(Node node)` to handle the unlinking logic.